docs/source/reference/types.rst

```diff
@@ -0,0 +1,286 @@
+
+.. code:: idris
+
+    [semiList] {a : Type} -> Semigroup (List a) where
+        (<+>) = (++)
+
+and
+
+.. code:: idris
+
+    semiList : {a : Type} -> Semigroup (List a)
+    semiList = (MkSemigroup (++))
+
+----------------------
+Other Forms
+----------------------
+
+The two main ways to declare types at the top level are ``record`` and ``data``
+
+.. note::
+    Type synonyms, type families, and the like from Haskell are all just normal functions from ``Type`` to something else.
+    In addition, this means that one can have anonymous instances, so, for instance, ``Monad (\x => x)`` is a perfectly valid form.
+
+    Even so, it is in general not a good idea to create such instances
+
+``data`` declares (General) Algebraic Data Types ((G)ADTs).
+The Haskell style, simple ADT form is always a statement of the form ``data NAME ARGS = ALT {| ALT}``.
+``ARGS`` must be a valid list of names.
+In addition, each ``ALT`` must be of the form ``CONS PARAMS`` where each ``CONS`` is either an operator or a name (that will be declared as a constructor).
+Finally, each of ``PARAMS`` must be a valid type.
+
+To declare (Agda style) Generalized Algebraic Data Types, we must write ``data NAME : KIND where CONS``.
+``KIND`` must be a kind (in Idris there isn't actually distinction between types and kinds) whose tail ultimately ends up returning the type ``Type``.
+So, for instance, ``data Vect : (n : Nat) -> Type where ...`` is valid, whilst ``data Vect : forall k. (n : Nat) -> k where ...`` is not.
+Each ``CONS`` must be a function declaration, except that the name being defined must start with an uppercase letter and the ultimate return type of it must be a fully applied instance of ``NAME``.
+
+The other way to define top level types, ``record``, is documented in detail on the `records`_ page.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Existentials in Types
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+While not restricted to types, existentials are so prevalent within types that there usage bears mentioning.
+Existentials are always of the form ``_``, ``?`` or ``?NAME``, where ``NAME`` is a valid name.
+Named existentials are scoped by *module*, however, they may only exist in one term.
+
+.. caution::
+
+    Existentials are scoped over modules.
+    This means that a given hole must have a unique name in a module.
+
+Named existential forms an equivalent to "typed holes", that is, a value to be filled in later.
+These can occur in types, although typically occur in generated ones.
+Most often, this will occur in usages of generic types, where, for instance, given ``show : {a : Type} -> Show a => a -> String``, we transform the context ``show : {a : Type} -> Show a => a -> String, y = show x |- y : String`` to ``x : ?a, Show ?a |- y : String`` [#f2]_ .
+
+``?`` is similar except, while if ``?a``` cannot be inferred it is made it a typed hole, if ``?`` cannot be inferred, it raises an error
```

Comment:
```suggestion
```

`?` is completely different as it infers a value in addition to the type. Holes only infer the type.